### ISSUE TITLE
Any way to revert to pre-7.0 isolated assembly loading in C++/CLI?

### ISSUE THREAD

**Issue by mikeoliphant** [dotnet/runtime#104480]
I have a framework using C++/CLI to load managed plugins within a native application.

This change to loading assemblies in the default ALC:

 https://github.com/dotnet/runtime/pull/66486

has broken my ability to load multiple plugins.

Is there any way to revert to the previous isolated assembly loading behavior?

**Comment by mikeoliphant** [comment]
To add a bit more information, what I have is:

- A native application host that loads native dll plugins (a [VST3](https://en.wikipedia.org/wiki/Virtual_Studio_Technology) audio plugin host, to be specific).
- Managed plugins wrapped in a C++/CLI wrapper framework to act as the native interface with the native host.

So, a "plugin" consists of a folder with the C++/CLI wrapper dll which loads the actual managed plugin code.

I can load a single plugin just fine. I can also load the same plugin multiple times from the same folder. Trying to load a second, different plugin, or even the same plugin, but from a different folder, no longer works (it does in .NET 6.0).

**Comment by AaronRobinsonMSFT** [comment]
@mikeoliphant There is no current way to revert to the previous C++/CLI behavior.

It would be helpful to get a trace - see instructions [here](https://learn.microsoft.com/dotnet/core/dependency-loading/collect-details).

/cc @elinor-fung

**Comment by mikeoliphant** [comment]
Here is a trace:

[vsthost.exe_20240705_181811.zip](https://github.com/user-attachments/files/16114570/vsthost.exe_20240705_181811.zip)

The application activity was:

- load a plugin
- load the same plugin again from the same location
- load the same plugin again, but from a different path

When the plugin is loaded from a different path, everything goes haywire and the trace keeps getting bigger and bigger. The one I linked above I cut off quickly to keep it from getting too big.

From looking at it with perfview (my first time with the tool, so I don't really know what I'm doing) it looks like it is loading the assembly "AudioPlugSharpVst" over and over again. That assembly is my C++/CLI assembly which is the entry point from the native code.

**Comment by AaronRobinsonMSFT** [comment]
@mikeoliphant Yep, I see the behavior you are referring to. The solution here is to have a different C++/CLI assembly. We changed this behavior as it more aligned with the majority of C++/CLI use cases coming from .NET Framework. Is it possible for you to adapt to the new behavior?

**Comment by mikeoliphant** [comment]
This is my github repo:

https://github.com/mikeoliphant/AudioPlugSharp

The C++/CLI assembly is a framework for loading managed plugins from native code (specifically, VST3 audio plugins).

If I'm understanding you correctly, I don't see how it would work for me to have different C++/CLI assemblies per plugin.

**Comment by AaronRobinsonMSFT** [comment]
Oh, I see. You provide a single C++/CLI assembly and anyone can drop that next to their plug-in written in .NET and it will get loaded, right?

**Comment by mikeoliphant** [comment]
> Oh, I see. You provide a single C++/CLI assembly and anyone can drop that next to their plug-in written in .NET and it will get loaded, right?

Yep, that's exactly it.

**Comment by AaronRobinsonMSFT** [comment]
I see. This is going to require feedback from @elinor-fung. My gut tells me this isn't something we are likely to add support for, but I'll defer to the hosting experts.

**Comment by mikeoliphant** [comment]
> My gut tells me this isn't something we are likely to add support for, but I'll defer to the hosting experts.

If so, that would be very disappointing if I have no possible way to work around it.

Can you give me some insight as to why it isn't working? My code *should* work in both the case that the separate instances of the C++/CLI assembly are loaded independently and isolated, and also in the case that only one copy is loaded and used.

**Comment by mikeoliphant** [comment]
Looking at this more closely, I don't think that it is even getting to my code when trying to load a plugin a second time from a different location.

Looking at the trace, I think you can see it succeeding the first two times (both loading from the same place) at around 1:18:18 and 1:18:24.

Then I try to load it from a different path (around 1:18:39) and it doesn't get past loading the main C++/CLI assembly ("AudioPlugSharpVst"). My actual plugin loading code is in a referenced assembly - "AudioPlugSharp" - and I don't think it even gets there. It just keeps loading "AudioPluginSharpVst" over and over again.

**Comment by mikeoliphant** [comment]
> I think a switch to allow the isolated ALC behaviour may be the best option here.

That should solve my problem, and would potentially be useful to others.

On the other hand, if, somehow, the behavior could be fixed so that the second assembly load works, but is in a the default ALC, I think I *might* be able make things work by getting the assembly location from the native side rather than the managed side.

I also *might* be able to work around the currently existing behavior by having another C++ dll as my entry point that always loads the AudioPlugSharpVst dll from a fixed location.

**Comment by mikeoliphant** [comment]
Please let me know if there is anything I can do to help move this forward.

**Comment by mikeoliphant** [comment]
Thanks very much for the detailed information - it is very helpful.

**Comment by mikeoliphant** [comment]
@elinor-fung Your instructions are clear, and the changes seem fairly straightforward.

I've cloned the runtime repository and built it. My struggle right now is with getting an environment set up to test changes.

I've got a Core_Root folder created, but I don't know how to make a plugin reference it. I don't think I can use corerun - it seems designed to work with managed assemblies?

I can put my modified ijwhost dll in my plugin folder, and that works, but I don't know how to test that alongside code changes to InMemoryAssemblyLoader.

Any help would be much appreciated.

**Comment by mikeoliphant** [comment]
Another issue I'm having is that I can't select .NET 9.0 as the target framework for my C++/CLI project.

I'm running latest preview version of Visual Studio, and have the 9.0.100-preview.6.24328.19 SDK installed.

**Comment by mikeoliphant** [comment]
I still haven't figured out how to test my plugin code against a dev version of the runtime, but I think I've successfully made the necessary changes to the runtime.

You can see what I've done on this branch:

https://github.com/mikeoliphant/runtime/tree/ijwhost_isolated_switch

There are three commits - one for the InMemoryAssemblyLoader change, one for the ijwhost changes and one for the test.

**Comment by mikeoliphant** [comment]
> I generally use the `libs.pretest` subset to create a a folder with my local build and set the `DOTNET_ROOT` environment variable when running an application to use it:

Thanks - that worked.

I've now verified that my plugin code works with my runtime changes.

As expected, it works when I add "System.Runtime.InteropServices.IJWHost.LoadComponentInIsolatedContext": true". If that config flag is "false" or absent, I get the previous behavior (assembly load loop when loading a plugin for a second time from a different location).

I think it is all good from my end. Let me know if you have any comments on the changes (for example, on the naming of the config option). Once I get a thumbs-up from you, I'll go ahead and submit a PR.

**Comment by mikeoliphant** [comment]
> * Config name: I'd recommend `IJWHost` -> `CppCLI` - so `System.Runtime.InteropServices.CppCLI.LoadComponentInIsolatedContext`. I think that better matches how we use C++/CLI as the feature name publicly (instead of IJW).
> * Tests: we should have the case of having a config with that property explicitly set to `false`. I'd find it clearer to add a separate `LoadLibrary_ContextConfig(bool loadIsolated)` test that always explicitly sets the property to true/false. And then leave the existing `LoadLibrary` test as validating the default behaviour (config property not set)

Ok - changes made:

https://github.com/dotnet/runtime/compare/main...mikeoliphant:runtime:ijwhost_isolated_switch

**Comment by mikeoliphant** [comment]
> For the test, it should still always add the property, but set it to `"true"` or `"false"` based on the value of `load_isolated` 

Fixed. PR is here: https://github.com/dotnet/runtime/pull/105337

Thanks for your help!
